csvy/validators/table_schema.py

```diff
@@ -0,0 +1,454 @@
+        + "['false', 'False', 'FALSE', '0'] is used.",
+    )
+
+
+class DateTimeColumnValidator(CommonValidator):
+    """Validator for the datetime columns in the Table Schema.
+
+    This class is used to validate the datetime columns in the Table Schema. It is based
+    on the columns defined in the Table Schema specification.
+
+    Attributes:
+        type_: A string specifying the type, with valid values being 'datetime'.
+        format_: A string specifying a format. It can be 'default' or 'any', or a string
+            with a custom format parsable by Python / C standard 'strptime', eg.
+            '%d/%m/%y'.
+
+    """
+
+    type_: Literal[TypeEnum.DATETIME, TypeEnum.DATE, TypeEnum.TIME] = Field(
+        TypeEnum.DATETIME, alias="type", description="A string specifying the type."
+    )
+    format_: FormatDateTime | str | None = Field(
+        None,
+        alias="format",
+        description="A string specifying a format. It can be "
+        + "'default' or 'any', or a string with a custom format parsable by Python / C "
+        + "standard 'strptime', eg. '%d/%m/%y'.",
+    )
+
+
+class GeopointColumnValidator(CommonValidator):
+    """Validator for the geopoint columns in the Table Schema.
+
+    This class is used to validate the geopoint columns in the Table Schema. It is based
+    on the columns defined in the Table Schema specification.
+
+    Attributes:
+        type_: A string specifying the type, with valid values being 'geopoint'.
+        format_: A string specifying a format, with valid values being 'default',
+            'array', 'object' and None.
+
+    """
+
+    type_: Literal[TypeEnum.GEOPOINT] = Field(
+        TypeEnum.GEOPOINT, alias="type", description="A string specifying the type."
+    )
+    format_: FormatGeoPoint | None = Field(
+        None, alias="format", description="A string specifying a format."
+    )
+
+
+class GeoJSONColumnValidator(CommonValidator):
+    """Validator for the geojson columns in the Table Schema.
+
+    This class is used to validate the geojson columns in the Table Schema. It is based
+    on the columns defined in the Table Schema specification.
+
+    Attributes:
+        type_: A string specifying the type, with valid values being 'geojson'.
+        format_: A string specifying a format, with valid values being 'default',
+            'topojson' and None.
+
+    """
+
+    type_: Literal[TypeEnum.GEOJSON] = Field(
+        TypeEnum.GEOJSON, alias="type", description="A string specifying the type."
+    )
+    format_: FormatGeoJSON | None = Field(
+        None, alias="format", description="A string specifying a format."
+    )
+
+
+ColumnValidator = Annotated[
+    StringColumnValidator
+    | NumberColumnValidator
+    | IntegerColumnValidator
+    | BooleanColumnValidator
+    | DateTimeColumnValidator
+    | GeopointColumnValidator
+    | GeoJSONColumnValidator,
+    Field(discriminator="type_"),
+]
+"""Annotated type for the ColumnValidator."""
```

Comment:
Can you explain what this is doing? What do we gain by using `Annotated` here? And what is `discriminator`?

Reply:
Ah, this is a trick I learn't from [PyECN config](https://github.com/ImperialCollegeLondon/LiECN/blob/055e5c9ce876cb934b30918f0e1b430415b03beb/pyecn/config/config.py#L158). The columns might have different formats (depending on their type) and each of them requires its own validation. We need a way to tell `pydantic` what column validator to use when parsing the input. This is the way to do it. Essentially it says:

> The column can be any of these options. Check the field `type_` and based on that decide which validator to use for the column.

Is it a bit clearer now?